File: sdk/python/atb/workflow_common.py
"""Shared helpers for profile workflow SDK recorders."""

from __future__ import annotations

import hashlib
import uuid
from collections.abc import Mapping
from datetime import UTC, datetime
from typing import Any, Literal

from atb.bundle import Bundle
from atb.canonicalize import canonicalize

DEFAULT_WORKFLOW_SAVE_PATH = "run.atb/bundle.atb"


class WorkflowContext:
    """Append profile events with optional auto-save."""

    def __init__(
        self,
        bundle: Bundle | None = None,
        *,
        auto_save: bool = False,
        save_path: str | None = None,
        actor_id: str | None = None,
        org_id: str | None = None,
        workspace_id: str | None = None,
    ) -> None:
        self.bundle = bundle if bundle is not None else Bundle()
        self.auto_save = auto_save
        self.save_path = save_path
        self.actor_id = actor_id
        self.org_id = org_id
        self.workspace_id = workspace_id
        self._request_bootstrapped = False

    def emit(self, event_type: str, payload: Mapping[str, Any]) -> None:
        self.bundle.append(
            event_type,
            dict(payload),
            actor_id=self.actor_id,
            org_id=self.org_id,
            workspace_id=self.workspace_id,
            timestamp=_now_rfc3339(),
        )
        if self.auto_save:
            self.bundle.save(self.save_path)

    def bootstrap_request(self, purpose_tag: str, request_id: str | None = None) -> str:
        rid = (request_id or "").strip() or f"req_{uuid.uuid4().hex}"
        if not self._request_bootstrapped:
            self.emit(
                "ai.request.received",
                {
                    "request_id": rid,
                    "actor_id_hash": actor_id_hash(self.actor_id),
                    "purpose_tag": purpose_tag,
                },
            )
            self._request_bootstrapped = True
        return rid

    def policy_payload(
        self,
        action_id: str,
        decision: Mapping[str, Any],
        subject_id: str | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "decision_id": action_id,
            "action_id": action_id,
            "policy_id": decision.get("policy_id", "local.workflow"),
            "policy_version": decision.get("policy_version", "v1"),
            "decision": decision["decision"],
            "decision_reason_codes": list(decision.get("reason_codes", [])),
        }
        subject_id_hash = subject_hash(subject_id, self.actor_id)
        if subject_id_hash is not None:
            payload["subject_id_hash"] = subject_id_hash
        return payload


def new_action_id(action_id: str | None = None) -> str:
    if action_id is not None and action_id.strip() != "":
        return action_id.strip()
    return f"act_{uuid.uuid4().hex}"


def new_job_id(job_id: str | None = None) -> str:
    if job_id is not None and job_id.strip() != "":
        return job_id.strip()
    return f"job_{uuid.uuid4().hex}"


def new_approval_id(approval_id: str | None = None) -> str:
    if approval_id is not None and approval_id.strip() != "":
        return approval_id.strip()
    return f"appr_{uuid.uuid4().hex}"


def canonical_digest(value: Any) -> str:
    return _sha256(canonicalize(value))


def value_digest(value: Any) -> str:
    try:
        encoded = canonicalize(value)
    except (TypeError, ValueError):
        encoded = repr(value).encode("utf-8")
    return _sha256(encoded)


def subject_hash(subject_id: str | None, actor_id: str | None) -> str | None:
    candidate = _normalize(subject_id)
    if candidate is None:
        candidate = _normalize(actor_id)
    if candidate is None:
        return None
    return _sha256(candidate.encode("utf-8"))


def actor_id_hash(actor_id: str | None) -> str:
    return subject_hash(None, actor_id) or _sha256(b"unknown")


def _sha256(value: bytes) -> str:
    return "sha256:" + hashlib.sha256(value).hexdigest()


def _normalize(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None


def now_rfc3339() -> str:
    return _now_rfc3339()


def _now_rfc3339() -> str:
    return datetime.now(UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")

File: sdk/typescript/src/workflow-common.ts
import { createHash, randomUUID } from "node:crypto";
import { Bundle } from "./bundle.js";
import { canonicalize } from "./canonicalize.js";
import { normalizeOptionalIdentity } from "./event.js";

export const DEFAULT_WORKFLOW_SAVE_PATH = "run.atb/bundle.atb";

/** Optional sink that receives workflow events instead of local bundle I/O. */
export interface WorkflowEventSink {
  append(eventType: string, payload: Record<string, unknown>): void;
}

/** Shared emitter options for profile workflow helpers. */
export interface WorkflowContextOptions {
  bundle?: Bundle;
  autoSave?: boolean;
  savePath?: string;
  actorId?: string;
  orgId?: string;
  workspaceId?: string;
  /** When set, events are sent to the local ATB Agent instead of the bundle file. */
  eventSink?: WorkflowEventSink;
}

/** Policy decision shape shared by workflow gates. */
export interface WorkflowPolicyDecision {
  decision: "allow" | "deny";
  reasonCodes?: string[];
  policyId?: string;
  policyVersion?: string;
}

/** Appends profile events with optional auto-save. */
export class WorkflowContext {
  readonly bundle: Bundle;

  private readonly autoSave: boolean;
  private readonly savePath: string;
  private readonly actorId?: string;
  private readonly orgId?: string;
  private readonly workspaceId?: string;
  private readonly eventSink?: WorkflowEventSink;
  private requestBootstrapped = false;

  constructor(options: WorkflowContextOptions = {}) {
    this.bundle = options.bundle ?? new Bundle();
    this.autoSave = options.autoSave ?? false;
    this.savePath = options.savePath ?? DEFAULT_WORKFLOW_SAVE_PATH;
    this.actorId = options.actorId;
    this.orgId = options.orgId;
    this.workspaceId = options.workspaceId;
    this.eventSink = options.eventSink;
  }

  /** Emit one event to the bundle. */
  emit(eventType: string, payload: Record<string, unknown>): void {
    if (this.eventSink) {
      this.eventSink.append(eventType, payload);
      return;
    }
    this.bundle.append(eventType, payload, {
      actorId: this.actorId,
      orgId: this.orgId,
      workspaceId: this.workspaceId,
      timestamp: nowRFC3339(),
    });
    if (this.autoSave) {
      this.bundle.save(this.savePath);
    }
  }

  /** Emit `ai.request.received` once per context when absent. */
  bootstrapRequest(purposeTag: string, requestId?: string): string {
    const rid = requestId?.trim() || `req_${randomUUID().replace(/-/g, "")}`;
    if (!this.requestBootstrapped) {
      this.emit("ai.request.received", {
        request_id: rid,
        actor_id_hash: actorIdHash(this.actorId),
        purpose_tag: purposeTag,
      });
      this.requestBootstrapped = true;
    }
    return rid;
  }

  /** Build a normalized policy decision payload. */
  policyPayload(
    actionId: string,
    decision: WorkflowPolicyDecision,
    subjectId?: string
  ): Record<string, unknown> {
    const normalized = normalizeWorkflowDecision(decision);
    const payload: Record<string, unknown> = {
      decision_id: actionId,
      action_id: actionId,
      policy_id: normalized.policyId,
      policy_version: normalized.policyVersion,
      decision: normalized.decision,
      decision_reason_codes: normalized.reasonCodes,
    };
    const subjectIdHash = subjectHash(subjectId, this.actorId);
    if (subjectIdHash !== undefined) {
      payload.subject_id_hash = subjectIdHash;
    }
    return payload;
  }
}

export function newActionId(actionId?: string): string {
  if (actionId && actionId.trim() !== "") {
    return actionId.trim();
  }
  return `act_${randomUUID().replace(/-/g, "")}`;
}

export function newJobId(jobId?: string): string {
  if (jobId && jobId.trim() !== "") {
    return jobId.trim();
  }
  return `job_${randomUUID().replace(/-/g, "")}`;
}

export function newApprovalId(approvalId?: string): string {
  if (approvalId && approvalId.trim() !== "") {
    return approvalId.trim();
  }
  return `appr_${randomUUID().replace(/-/g, "")}`;
}

export function canonicalDigest(value: unknown): string {
  return sha256(Buffer.from(canonicalize(value), "utf8"));
}

export function valueDigest(value: unknown): string {
  try {
    return sha256(Buffer.from(canonicalize(value), "utf8"));
  } catch {
    return sha256(Buffer.from(String(value), "utf8"));
  }
}

export function subjectHash(
  subjectId: string | undefined,
  actorId: string | undefined
): string | undefined {
  const candidate =
    normalizeOptionalIdentity(subjectId) ?? normalizeOptionalIdentity(actorId);
  if (candidate === undefined) {
    return undefined;
  }
  return sha256(Buffer.from(candidate, "utf8"));
}

export function actorIdHash(actorId: string | undefined): string {
  return subjectHash(undefined, actorId) ?? sha256(Buffer.from("unknown", "utf8"));
}

export function normalizeWorkflowDecision(
  decision: WorkflowPolicyDecision
): Required<WorkflowPolicyDecision> {
  return {
    decision: decision.decision,
    reasonCodes: decision.reasonCodes ?? [],
    policyId: decision.policyId ?? "local.workflow",
    policyVersion: decision.policyVersion ?? "v1",
  };
}

function sha256(value: Uint8Array): string {
  return `sha256:${createHash("sha256").update(value).digest("hex")}`;
}

function nowRFC3339(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}
